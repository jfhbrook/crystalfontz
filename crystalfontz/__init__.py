# -*- coding: utf-8 -*-

"""
This is the documentation for crystalfontz.
"""

