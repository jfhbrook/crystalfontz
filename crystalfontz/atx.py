from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Optional, Self

AUTO_POLARITY = 0x01


class AtxPowerSwitchFunction(Enum):
    RESET_INVERT = 0x02
    POWER_INVERT = 0x04
    LCD_OFF_IF_HOST_IS_OFF = 0x10
    KEYPAD_RESET = 0x20
    KEYPAD_POWER_ON = 0x40
    KEYPAD_POWER_OFF = 0x80


@dataclass
class AtxPowerSwitchFunctionalitySettings:
    functions: Iterable[AtxPowerSwitchFunction]
    auto_polarity: bool = True
    power_pulse_length_seconds: Optional[float] = None

    def to_bytes(self: Self) -> bytes:
        functions: int = 0
        for function in self.functions:
            functions ^= function.value
        if self.auto_polarity:
            functions ^= AUTO_POLARITY

        packed: bytes = functions.to_bytes()

        if self.power_pulse_length_seconds is not None:
            pulse_length = int(self.power_pulse_length_seconds * 32)

            if pulse_length < 1:
                raise ValueError(f"Pulse length can not be less than {1/32}")

            packed += min(pulse_length, 255).to_bytes()

        return packed
