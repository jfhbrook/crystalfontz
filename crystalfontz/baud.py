from typing import Literal

BaudRate = Literal[19200] | Literal[115200]
