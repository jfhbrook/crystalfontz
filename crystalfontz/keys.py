from dataclasses import asdict, dataclass
from enum import Enum
from typing import List, Literal, Self, Type

KeyPress = (
    Literal[0x01]
    | Literal[0x02]
    | Literal[0x04]
    | Literal[0x08]
    | Literal[0x10]
    | Literal[0x20]
)

KP_UP: KeyPress = 0x01
KP_ENTER: KeyPress = 0x02
KP_EXIT: KeyPress = 0x04
KP_LEFT: KeyPress = 0x08
KP_RIGHT: KeyPress = 0x10
KP_DOWN: KeyPress = 0x20


@dataclass
class KeyState:
    """
    A key's state.

    Attributes:
        pressed (bool): When True, the key is currently pressed.
        pressed_since (bool): When True, the key has been pressed since the last poll.
        released_since (bool): When True, the key has been released since the last
                               poll.
    """

    pressed: bool
    pressed_since: bool
    released_since: bool

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes, keypress: KeyPress) -> Self:
        pressed = state[0]
        pressed_since = state[1]
        released_since = state[2]

        return cls(
            pressed=bool(pressed & keypress),
            pressed_since=bool(pressed_since & keypress),
            released_since=bool(released_since & keypress),
        )

    def to_bytes(self: Self, keypress: KeyPress) -> bytes: ...


@dataclass
class KeyStates:
    """
    The state of all keys.

    Attributes:
        up: The state of the "up" key.
        enter: The state of the "enter" key.
        exit: The state of the "exit" key.
        left: The state of the "left" key.
        right: The state of the "right" key.
        down: The state of the "down" key.
    """

    up: KeyState
    enter: KeyState
    exit: KeyState
    left: KeyState
    right: KeyState
    down: KeyState

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes) -> Self:
        return cls(
            up=KeyState.from_bytes(state, KP_UP),
            enter=KeyState.from_bytes(state, KP_ENTER),
            exit=KeyState.from_bytes(state, KP_EXIT),
            left=KeyState.from_bytes(state, KP_LEFT),
            right=KeyState.from_bytes(state, KP_RIGHT),
            down=KeyState.from_bytes(state, KP_DOWN),
        )

    def to_bytes(self: Self) -> bytes:
        pressed = 0x00
        pressed_since = 0x00
        released_since = 0x00

        for state, keypress in [
            (self.up, KP_UP),
            (self.enter, KP_ENTER),
            (self.exit, KP_EXIT),
            (self.left, KP_LEFT),
            (self.right, KP_RIGHT),
            (self.down, KP_DOWN),
        ]:
            pressed = (pressed ^ keypress) if state.pressed else pressed
            pressed_since = (
                (pressed_since ^ keypress) if state.pressed_since else pressed_since
            )
            released_since = (
                (released_since ^ keypress) if state.released_since else released_since
            )

        return bytes([pressed, pressed_since, released_since])

    def __repr__(self: Self) -> str:
        repr_ = ""
        for name, state in asdict(self).items():
            st = ", ".join([f"{n}={'yes' if s else 'no'}" for n, s in state.items()])
            repr_ += f"{name}: {st}\n"
        return repr_[0:-1]


class KeyActivity(Enum):
    """
    A key activity. This is either a "press" event or a "release" event for a
    given key.
    """

    KEY_UP_PRESS = 1
    KEY_DOWN_PRESS = 2
    KEY_LEFT_PRESS = 3
    KEY_RIGHT_PRESS = 4
    KEY_ENTER_PRESS = 5
    KEY_EXIT_PRESS = 6
    KEY_UP_RELEASE = 7
    KEY_DOWN_RELEASE = 8
    KEY_LEFT_RELEASE = 9
    KEY_RIGHT_RELEASE = 10
    KEY_ENTER_RELEASE = 11
    KEY_EXIT_RELEASE = 12

    @classmethod
    def from_bytes(cls: Type[Self], activity: bytes) -> "KeyActivity":
        return KEY_ACTIVITIES[activity[0] - 1]


KEY_ACTIVITIES: List[KeyActivity] = [
    KeyActivity.KEY_UP_PRESS,
    KeyActivity.KEY_DOWN_PRESS,
    KeyActivity.KEY_LEFT_PRESS,
    KeyActivity.KEY_RIGHT_PRESS,
    KeyActivity.KEY_ENTER_PRESS,
    KeyActivity.KEY_EXIT_PRESS,
    KeyActivity.KEY_UP_RELEASE,
    KeyActivity.KEY_DOWN_RELEASE,
    KeyActivity.KEY_LEFT_RELEASE,
    KeyActivity.KEY_RIGHT_RELEASE,
    KeyActivity.KEY_ENTER_RELEASE,
    KeyActivity.KEY_EXIT_RELEASE,
]
