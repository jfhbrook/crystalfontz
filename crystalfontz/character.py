from typing import Dict, Self

from crystalfontz.error import EncodeError

"""
Abstractions for representing character ROMs (CGROMs), and encoding text to be
used with them.
"""

# Characters that take more than one code point in unicode. These need to be
# manually added to the character ROM.
super_minus: str = "\u207b"
super_one: str = "\u00B9"
inverse: str = f"{super_minus}{super_one}"
x_bar: str = "x̄"

# Characters which only require one code point, but are difficult to type
# and/or are ambiguous.
hbar = "―"  # Intended to be a katakana character?
block = "█"

# Japanese punctuation. The CFA533, at least, contains Katakana characters
# with corresponding punctuation.
japan_interpunct = "・"
japan_lquote = "「"
japan_rquote = "」"
japan_full_stop = "。"
japan_comma = "、"


class CgRom:
    def __init__(self: Self, sheet: str) -> None:
        self._table: Dict[str, bytes] = dict()

        lines = sheet.split("\n")
        if lines[0] == "":
            lines = lines[1:]
        if lines[-1] == "":
            lines = lines[0:-1]

        for i, row in enumerate(lines):
            for j, char in enumerate(row):
                point = (16 * j) + i
                if char != " " or point == 32:
                    self._table[char] = point.to_bytes()

    def __getitem__(self: Self, key: str) -> bytes:
        return self._table[key]

    def __setitem__(self: Self, key: str, value: bytes) -> None:
        self._table[key] = value

    def set_encoding(self: Self, char: str, encoded: int | bytes) -> Self:
        if isinstance(encoded, int):
            self[char] = encoded.to_bytes()
        else:
            self[char] = encoded
        return self

    def encode(self: Self, input: str, errors="strict") -> bytes:
        output: bytes = b""
        i = 0
        while i < len(input):
            char = input[i]

            # TODO: This encoder uses if/else statements to handle multi-byte
            # encodings. To make this general purpose, it needs to be
            # refactored to use a trie-like structure.
            if char == "x":
                if input[i + 1] == x_bar[1] and input[i + 2] == x_bar[2]:
                    output += self._table[x_bar]
                    i += 2
                else:
                    output += self._table["x"]
            elif char in self._table:
                output += self._table[char]
            elif char == super_minus:
                if input[i + 1] == super_one:
                    output += self._table[inverse]
                    i += 1
            else:
                if errors == "strict":
                    raise EncodeError(f"Unknown character {char}")
                else:
                    output += self._table["*"]

            i += 1

        return output


#
# This ROM encoding is based on page 44 of this doc:
#
#     file:///Users/josh/Downloads/CFA533-TMI-KU.pdf
#
# However, it is *incomplete*, mostly because I don't know katakana and only
# know a smattering of Greek. Unknown characters are filled in with spaces.
# Some characters that *are* filled out are best guesses.
#
# NOTE: ASCII characters generally share their code points with true ASCII.
# TODO: Does this ROM match another encoding which contains both katakana and
# Greek letters?
# NOTE: The first column in the ROM is reserved for custom characters.
#

ROM = (
    CgRom(
        """
   0@P`p   ―  α 
  !1AQaq  。ア  ä 
  "2BRbr  「 ツ βθ
  #3CScs  」ウ  ε∞
  $4DTdt  、エ  μΩ
  %5EUeu  ・オ  σü
  &6FVfv  ヲカ  ρΣ
  '7GWgw  アキ   π
  (8HXhx  イク  √ 
  )9IYiy  ゥケ    
  *:JZjz  エコ     
  +;K[k{  オサ    
  ,<L¥l|  ヤシ  ¢ 
  -=M]m}  ユヌ  £÷
  .>N^n→  ヨセ  ñ 
  /?O_o←  ツソ °ö█
"""
    )
    .set_encoding(inverse, 244 + 9)
    .set_encoding(x_bar, 240 + 8)
)


# TODO: This encoding is likely device specific. I'm leaving it in here for
# now because the encode method contains logic specific to the device. However,
# once that is factored out in favor of a trie-like data structure, this
# specific instance should be moved to device.py.
encode_chars = ROM.encode
