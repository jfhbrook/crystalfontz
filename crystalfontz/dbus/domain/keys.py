from typing import ClassVar, Set, Tuple

from crystalfontz.dbus.domain.base import ByteM, OptFloatM, OptFloatT, struct
from crystalfontz.keys import (
    KeyPress,
    KeyState,
    KeyStates,
    KP_DOWN,
    KP_ENTER,
    KP_EXIT,
    KP_LEFT,
    KP_RIGHT,
    KP_UP,
)

KeyPressT = int

KEYPRESSES: Set[KeyPress] = {KP_UP, KP_ENTER, KP_EXIT, KP_LEFT, KP_RIGHT, KP_DOWN}


class KeyPressM(ByteM):
    t: ClassVar[str] = ByteM.t

    @staticmethod
    def unpack(keypress: int) -> KeyPress:
        if keypress not in KEYPRESSES:
            raise ValueError(f"{keypress} is not a valid key press")
        return keypress


KeyStateT = Tuple[bool, bool, bool]
KeyStatesT = Tuple[
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
]


class KeyStateM:
    t: ClassVar[str] = struct("bbb")

    @staticmethod
    def pack(state: KeyState) -> KeyStateT:
        return (state.pressed, state.pressed_since, state.released_since)

    @staticmethod
    def unpack(state: KeyStateT) -> KeyState:
        pressed, pressed_since, released_since = state
        return KeyState(
            pressed=pressed, pressed_since=pressed_since, released_since=released_since
        )


class KeyStatesM:
    t: ClassVar[str] = struct(KeyStateM.t) * 6

    @staticmethod
    def pack(states: KeyStates) -> KeyStatesT:
        return (
            KeyStateM.pack(states.up),
            KeyStateM.pack(states.enter),
            KeyStateM.pack(states.exit),
            KeyStateM.pack(states.left),
            KeyStateM.pack(states.right),
            KeyStateM.pack(states.down),
        )

    @staticmethod
    def unpack(states: KeyStatesT) -> KeyStates:
        up, enter, exit, left, right, down = states
        return KeyStates(
            up=KeyStateM.unpack(up),
            enter=KeyStateM.unpack(enter),
            exit=KeyStateM.unpack(exit),
            left=KeyStateM.unpack(left),
            right=KeyStateM.unpack(right),
            down=KeyStateM.unpack(down),
        )


KeypadBrightnessT = OptFloatT


class KeypadBrightnessM(OptFloatM):
    """
    Map `Optional[float]` to and from `KeypadBrightnessT` (`float`).

    `KeypadBrightnessM` is an alias for `OptFloatM`.
    """

    t: ClassVar[str] = OptFloatM.t
