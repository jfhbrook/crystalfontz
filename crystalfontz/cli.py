import asyncio
from dataclasses import dataclass
import functools
import logging
from typing import Callable, cast, Coroutine, Literal, Optional, Type

import click

from crystalfontz.baud import BaudRate, FAST_BAUD_RATE, SLOW_BAUD_RATE
from crystalfontz.client import Client, create_connection
from crystalfontz.config import Config, GLOBAL_FILE
from crystalfontz.report import NoopReportHandler, ReportHandler


@dataclass
class Obj:
    config: Config
    global_: bool
    port: str
    baud_rate: BaudRate


LogLevel = (
    Literal["DEBUG"]
    | Literal["INFO"]
    | Literal["WARNING"]
    | Literal["ERROR"]
    | Literal["CRITICAL"]
)


@click.group()
@click.option(
    "--global/--no-global",
    "global_",
    default=False,
    help="Load the global config file at {GLOBAL_FILE}",
)
@click.option(
    "--log-level",
    environ="CRYSTALFONTZ_LOG_LEVEL",
    type=click.Choice(["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]),
    default="INFO",
    help="Set the log level",
)
@click.option(
    "--port",
    environ="CRYSTALFONTZ_PORT",
    help="The serial port the Crystalfontz LCD is connected to",
)
@click.option(
    "--baud",
    type=click.Choice([str(SLOW_BAUD_RATE), str(FAST_BAUD_RATE)]),
    environ="CRYSTALFONTZ_BAUD_RATE",
    help="The baud rate to use when connecting to the Crystalfontz LCD",
)
@click.pass_context
def main(
    ctx: click.Context,
    global_: bool,
    log_level: LogLevel,
    port: Optional[str],
    baud: Optional[str],
) -> None:
    baud_rate = cast(Optional[BaudRate], int(baud) if baud else None)
    config: Config = Config.from_file(file=GLOBAL_FILE if global_ else None)
    ctx.obj = Obj(
        config=config,
        global_=global_,
        port=port or config.port,
        baud_rate=baud_rate or config.baud_rate,
    )

    logging.basicConfig(level=getattr(logging, log_level))


AsyncCommand = Callable[..., Coroutine[None, None, None]]
WrappedAsyncCommand = Callable[..., None]
AsyncCommandDecorator = Callable[[AsyncCommand], WrappedAsyncCommand]


def client(
    run_forever: bool = False,
    report_handler_cls: Type[ReportHandler] = NoopReportHandler,
) -> AsyncCommandDecorator:
    def decorator(fn: AsyncCommand) -> WrappedAsyncCommand:
        @click.pass_context
        @functools.wraps(fn)
        def wrapped(ctx: click.Context, *args, **kwargs) -> None:
            port: str = ctx.obj.port
            baud_rate: BaudRate = ctx.obj.baud_rate

            async def main() -> None:
                client: Client = await create_connection(
                    port, report_handler=report_handler_cls(), baud_rate=baud_rate
                )
                await fn(client, *args, **kwargs)

            try:
                if run_forever:
                    loop = asyncio.new_event_loop()
                    asyncio.set_event_loop(loop)
                    loop.create_task(main())
                    loop.run_forever()
                else:
                    asyncio.run(main())
            except KeyboardInterrupt:
                pass

        return wrapped

    return decorator


@main.command()
def listen() -> None:
    """
    Listen for key and temperature reports. To configure which reports to
    receive, use 'crystalfontz keypad configure-reporting' and
    'crystalfontz temperature setup-reporting' respectively.
    """
    pass


@main.command(help="0 (0x00): Ping command")
@click.argument("payload")
@client()
async def ping(client: Client, payload: str) -> None:
    await client.ping(payload.encode("utf8"))


@main.command()
def versions() -> None:
    pass


@main.group()
def user_flash_area() -> None:
    pass


@user_flash_area.command(name="write")
def write_user_flash_area() -> None:
    pass


@user_flash_area.command(name="read")
def read_user_flash_area() -> None:
    pass


@main.command()
def store_boot_state() -> None:
    pass


@main.group()
def lcd() -> None:
    pass


@lcd.command(name="reboot")
def reboot_lcd() -> None:
    pass


@main.group()
def host() -> None:
    pass


@host.command(name="reset")
def reset_host() -> None:
    pass


@host.command(name="shutdown")
def shutdown_host() -> None:
    pass


@main.command()
def clear_screen() -> None:
    pass


@main.command()
def set_line_1() -> None:
    pass


@main.command()
def set_line_2() -> None:
    pass


@main.group()
def special_character() -> None:
    pass


@special_character.command(name="set-data")
def set_special_character_data() -> None:
    pass


@special_character.command(name="set-encoding")
def set_special_character_encoding() -> None:
    pass


@lcd.command(name="read_memory")
def read_lcd_memory() -> None:
    pass


@main.group()
def cursor() -> None:
    pass


@cursor.command(name="set-position")
def set_cursor_position() -> None:
    pass


@cursor.command(name="set-style")
def set_cursor_style() -> None:
    pass


@main.command()
def set_contrast() -> None:
    pass


@main.command()
def set_backlight() -> None:
    pass


@main.group()
def dow() -> None:
    pass


@dow.command(name="read-device-information")
def read_dow_device_information() -> None:
    pass


@main.group()
def temperature() -> None:
    pass


@temperature.command(name="setup-reporting")
def setup_temperature_reporting() -> None:
    pass


@dow.command(name="transaction")
def dow_transaction() -> None:
    pass


@temperature.command(name="setup-live-display")
def setup_live_temperature_display() -> None:
    pass


@lcd.command(name="send-command")
def send_command_to_lcd_controler() -> None:
    pass


@main.group()
def keypad() -> None:
    pass


@keypad.command(name="poll")
def poll_keypad() -> None:
    pass


@main.group()
def atx_power_switch() -> None:
    pass


@atx_power_switch.command(name="set-functionality")
def set_atx_power_switch_functionality() -> None:
    pass


@main.command()
def configure_watchdog() -> None:
    pass


@main.command()
def status() -> None:
    pass


@main.command()
def send() -> None:
    pass


@main.command()
def set_baud_rate() -> None:
    pass


@main.group()
def gpio() -> None:
    pass


@gpio.command(name="set")
def set_gpio() -> None:
    pass


@gpio.command(name="read")
def read_gpio() -> None:
    pass
