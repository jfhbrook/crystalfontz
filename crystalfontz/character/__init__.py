from crystalfontz.character.constants import (
    block,
    hbar,
    inverse,
    japan_comma,
    japan_full_stop,
    japan_interpunct,
    japan_lquote,
    japan_rquote,
    x_bar,
)
from crystalfontz.character.rom import CharacterRom
from crystalfontz.character.special import SpecialCharacter

#
# This ROM encoding is based on page 44 of CFA533-TMI-KU.pdf.
#
# However, it is *incomplete*, mostly because I don't know katakana and only
# know a smattering of Greek. Unknown characters are filled in with spaces.
# Some characters that *are* filled out are best guesses.
#
# NOTE: ASCII characters generally share their code points with true ASCII.
# TODO: Does this ROM match another encoding which contains both katakana and
# Greek letters?
# NOTE: The first column in the ROM is reserved for custom characters.
#

CHARACTER_ROM = (
    CharacterRom(
        """
   0@P`p   ―  α 
  !1AQaq  。ア  ä 
  "2BRbr  「 ツ βθ
  #3CScs  」ウ  ε∞
  $4DTdt  、エ  μΩ
  %5EUeu  ・オ  σü
  &6FVfv  ヲカ  ρΣ
  '7GWgw  アキ   π
  (8HXhx  イク  √ 
  )9IYiy  ゥケ    
  *:JZjz  エコ     
  +;K[k{  オサ    
  ,<L¥l|  ヤシ  ¢ 
  -=M]m}  ユヌ  £÷
  .>N^n→  ヨセ  ñ 
  /?O_o←  ツソ °ö█
"""
    )
    .set_encoding(inverse, 244 + 9)
    .set_encoding(x_bar, 240 + 8)
)

# TODO: This encoding is likely device specific. I'm leaving it in here for
# now because the encode method contains logic specific to the device. However,
# once that is factored out in favor of a trie-like data structure, this
# specific instance should be moved to device.py.
encode_chars = CHARACTER_ROM.encode

__all__ = [
    "inverse",
    "x_bar",
    "hbar",
    "block",
    "japan_interpunct",
    "japan_lquote",
    "japan_rquote",
    "japan_full_stop",
    "japan_comma",
    "CharacterRom",
    "SpecialCharacter",
    "encode_chars",
]
