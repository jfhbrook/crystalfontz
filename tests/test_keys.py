from crystalfontz.keys import KeyState, KeyStates


def test_key_states_to_from_bytes(snapshot) -> None:
    key_states = KeyStates(
        up=KeyState(pressed=False, pressed_since=False, released_since=False),
        enter=KeyState(pressed=True, pressed_since=False, released_since=False),
        exit=KeyState(pressed=False, pressed_since=True, released_since=False),
        left=KeyState(pressed=False, pressed_since=False, released_since=True),
        right=KeyState(pressed=True, pressed_since=True, released_since=False),
        down=KeyState(pressed=True, pressed_since=False, released_since=True),
    )

    as_bytes = key_states.to_bytes()

    assert as_bytes == snapshot

    from_bytes = KeyStates.from_bytes(as_bytes)

    assert from_bytes == key_states
