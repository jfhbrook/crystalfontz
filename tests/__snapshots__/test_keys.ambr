# serializer version: 1
# name: test_key_states_to_from_bytes
  b'\x01\x00\x00'
# ---
